contrib/endpoints/src/api_manager/service_control/proto.cc

```diff
@@ -905,6 +911,83 @@ Proto::Proto(const std::set<std::string>& logs,
       service_name_(service_name),
       service_config_id_(service_config_id) {}
 
+utils::Status Proto::FillAllocateQuotaRequest(
+    const QuotaRequestInfo& info,
+    ::google::api::servicecontrol::v1::AllocateQuotaRequest* request) {
+  ::google::api::servicecontrol::v1::QuotaOperation* operation =
+      request->mutable_allocate_operation();
+
+  // service_name
+  request->set_service_name(service_name_);
+  // service_config_id
+  request->set_service_config_id(service_config_id_);
+
+  // allocate_operation.operation_id
+  if (!info.operation_id.empty()) {
+    operation->set_operation_id(info.operation_id);
+  }
+  // allocate_operation.method_name
+  if (!info.method_name.empty()) {
+    operation->set_method_name(info.method_name);
+  }
+  // allocate_operation.consumer_id
+  if (!info.api_key.empty()) {
+    operation->set_consumer_id(std::string(kConsumerIdApiKey) +
+                               std::string(info.api_key));
+  }
+
+  // TODO(jaebong) - Read from service conf?
+  // allocate_operation.quota_mode
+  operation->set_quota_mode(
+      ::google::api::servicecontrol::v1::QuotaOperation_QuotaMode::
+          QuotaOperation_QuotaMode_NORMAL);
+
+  // allocate_operation.labels
+  auto* labels = operation->mutable_labels();
+  if (!info.client_ip.empty()) {
+    (*labels)[kServiceControlCallerIp] = info.client_ip;
+  }
+
+  if (!info.referer.empty()) {
+    (*labels)[kServiceControlReferer] = info.referer;
+  }
+  (*labels)[kServiceControlUserAgent] = kUserAgent;
+  (*labels)[kServiceControlServiceAgent] =
+      kServiceAgentPrefix + utils::Version::instance().get();
+
+  // allocate_operation.quota_metrics
+  // TODO(jaebong) need to update after the configuration module change
+  // handle group rules
+  if (info.quota_rule_ != nullptr) {
```

Comment:
info will have a raw pointer to a 

const map<metric_name, metric_cost>*

you just use that to construct metrics.